README.md
# CinnamonBase

CinnamonBase is a collection of basic classes which are needed in all server related projects
of the Cinnamon CMS server, version 2.x.

License: LGPL 2.1 or later

To build this library, use the build script from Cinnamon server, which can be found at

* https://github.com/dewarim/cinnamon2-server/blob/master/build.xml

Homepage: http://cinnamon-cms.de
Sourceforge project page: http://sourceforge.net/projects/cinnamon

Maintainer: ingo.wiarda@horner-project.de


changelog.txt
2.4.0
+ Moved to new git repository.

2.0.2
+ Fixed: CinnamonException could in some circumstances (null parameter) fail to create a i18n message.

2.0.1
+ Added new Constants: all default index item and index field names.

1.0.9
+ Added "failed" state to procstate-constants.

src/utils/ParamParser.java
package utils;

import org.dom4j.Document;
import org.dom4j.DocumentException;
import org.dom4j.DocumentHelper;
import org.dom4j.Node;
import org.dom4j.io.OutputFormat;
import org.dom4j.io.SAXReader;
import org.dom4j.io.XMLWriter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.StringReader;
import java.io.StringWriter;
import java.util.Date;

public class ParamParser {

	public static Long parseLong(String param, String message){
		try {
			return Long.parseLong(param);
		} catch (NumberFormatException e) {
			throw new RuntimeException(message);
		}
	}
	public static Integer parseInt(String param, String message){
		try {
			return Integer.parseInt(param);
		} catch (NumberFormatException e) {
			throw new RuntimeException(message);
		}
	}

    /**
     * Parses a string to a boolean value. The parameter is converted to lower case before
     * parsing.
     * @param param the string to parse
     * @param message the message in case something goes wrong.
     * @return the boolean result.
     */
    public static Boolean parseBoolean(String param, String message){
        
        try{
            String p = param.trim().toLowerCase();
			if(p.equals("true")){
                return true;
            }
            else if(p.equals("false")){
                return false;
            }
            else{
                throw new RuntimeException("error.param");
            }
		} catch (RuntimeException e) {
			throw new RuntimeException(message);
		}
    }

	public static Node parseXml(String xml, String message){
		return ParamParser.parseXmlToDocument(xml, message).getRootElement().detach();	
	}

    public static Document parseXmlToDocument(String xml){
        return parseXmlToDocument(xml, null);
    }

	public static Document parseXmlToDocument(String xml, String message){
		if(message == null){
			 message = "error.parse.xml";
		}
		try{
			xml = xml.replaceAll("^(?:\\xEF\\xBB\\xBF|\uFEFF)", ""); // remove BOM on UTF-8 Strings.
			SAXReader reader = new SAXReader();
			// ignore dtd-declarations
			reader.setIncludeExternalDTDDeclarations(false);
			// do not validate - we are only interested in receiving a doc.
			reader.setValidation(false);
			return reader.read(new StringReader(xml));
		}
		catch (DocumentException e) {
            Logger log = LoggerFactory.getLogger(ParamParser.class);
            log.debug("ParamParser::DocumentException::", e);
			throw new RuntimeException(message);
		}
	}
	
	public static String dateToIsoString(Date date){
		return String.format("%1$tFT%1$tT", date);
	}
	
	// Could be placed in a more specific XML class.
	/**
	 * Turn a raw XML string and return it in an indented
	 * and more human readable form.
	 * @param xml an XML document as string.
	 * @return an indented version of the input XML.  
	 */
	public static String prettyPrint(String xml){
		String prettyXML;
		try {
			OutputFormat format = OutputFormat.createPrettyPrint();
			Document doc = DocumentHelper.parseText(xml);
			StringWriter sw = new StringWriter();
			XMLWriter xw = new XMLWriter(sw, format);
			xw.write(doc);
			prettyXML = sw.toString();
			
		} catch (Exception e) {
			throw new RuntimeException(e);
		}
		return prettyXML;
	}
}


src/utils/ContentReader.java
package utils;

import org.apache.commons.io.FileUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import server.exceptions.CinnamonException;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;

public class ContentReader {
    static Logger log = LoggerFactory.getLogger(ContentReader.class);

    public static String readFileAsString(String filePath)
            throws IOException {
        return readFileAsString(filePath, null);
    }

    public static String readFileAsString(String filePath, String encoding) throws IOException {
        if(filePath == null){
            throw new CinnamonException("error.path.is.null");
        }
        log.debug("reading from file: " + filePath);
        File f = new File(filePath);
        checkExistence(f);
        checkDirectory(f);
        return FileUtils.readFileToString(f, encoding);
    }

    public static byte[] readFileAsBytes(String filePath) throws IOException {
        if(filePath == null){
            throw new CinnamonException("error.path.is.null");
        }
        log.debug("reading from file: " + filePath);
        File f = new File(filePath);
        checkExistence(f);
        checkDirectory(f);
        return FileUtils.readFileToByteArray(f);
    }

    static void checkExistence(File file) throws FileNotFoundException {
        if (!file.exists()) {
            log.warn("File not found: " + file.getAbsolutePath());
            throw new FileNotFoundException("error.file_not_found");
        }
    }

    static void checkDirectory(File file) throws IOException {
        if (file.isDirectory()) {
            throw new IOException("error.filename.is_directory");
        }
    }

}


src/server/global/Constants.java
// cinnamon - the Open Enterprise CMS project
// Copyright (C) 2007-2009 Horner GmbH (http://www.horner-project.eu)
// 
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
// 
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
// (or visit: http://www.gnu.org/licenses/lgpl.html)

package server.global;

public class Constants {

    public static final String SERVER_VERSION = "2.4.1";
    
	public static final String PS_LABEL_CREATED="_created";
	public static final String ACL_DEFAULT="_default_acl";

	public static final String OBJTYPE_DEFAULT = "_default_objtype";
    public static final String OBJTYPE_TRANSLATION_TASK = "_translation_task";
	public static final String OBJTYPE_WORKFLOW = "_workflow";
	/**
	 * The name of the object type for workflow template objects.
	 */
	public static final String OBJTYPE_WORKFLOW_TEMPLATE = "_workflow_template";
	public static final String OBJTYPE_TASK_DEFINITION= "_task_definition";
	public static final String OBJTYPE_TASK = "_task";
	public static final String OBJTYPE_SEARCH = "_search";
	public static final String OBJTYPE_SUPPORTING_DOCUMENT = "_supporting_document";
	public static final String OBJTYPE_RENDITION = "_rendition";
	public static final String OBJTYPE_CART = "_cart";
	public static final String OBJTYPE_CONFIG = "_config";
	public static final String OBJTYPE_IMAGE = "image";
	public static final String OBJTYPE_DOCUMENT = "document";

	public static final String FOLDER_TYPE_DEFAULT = "_default_folder_type";
	public static final String INDEX_GROUP_DEFAULT = "_default_index_group";
	
	public static final String GROUP_SUPERUSERS="_superusers";
    public static final String GROUP_USERS = "_users";
	public static final String ALIAS_EVERYONE="_everyone";
	public static final String ALIAS_OWNER="_owner";
	
	public static final String ROOT_FOLDER_NAME = "root";
	public static final String SYSTEM_FOLDER_NAME = "system";
	public static final String WORKFLOW_FOLDER_NAME = "workflows";
	public static final String WORKFLOW_TEMPLATE_FOLDER_NAME = "templates";
	public static final String WORKFLOW_TASK_DEFINITION_FOLDER_NAME = "task_definitions";
	public static final String WORKFLOW_TASK_FOLDER_NAME = "tasks";

    public static final String RELATION_RESOLVER_FIXED = "FixedRelationResolver";
	public static final String RELATION_TYPE_WORKFLOW_TO_START_TASK = "_workflow_start_task";
	public static final String RELATION_TYPE_WORKFLOW_TO_TASK = "_workflow_task";
	public static final String RELATION_TYPE_WORKFLOW_TO_DEADLINE_TASK = "_workflow_deadline_task";
    public static final String RELATION_TYPE_CHILD = "child_content";
    public static final String RELATION_TYPE_RENDITION = "rendition";
    public static final String RELATION_TYPE_TRANSLATION_SOURCE = "translation_source";
    public static final String RELATION_TYPE_TRANSLATION_ROOT = "translation_root";
    public static final String RELATION_TYPE_CHILD_NO_CONTENT = "child_no_content";
    public static final String RELATION_TYPE_TRANSLATION_SOURCE_LIST = "translation_source_list";
    public static final String RELATION_TYPE_TRANSLATION_TARGET_LIST = "translation_target_list";

    public static final String USER_SUPERADMIN_NAME = "admin";

	public static final String WORKFLOW_FOLDER_PATH = "/root/system/workflows";
	public static final String WORKFLOW_TEMPLATE_PATH = WORKFLOW_FOLDER_PATH +"/"+ WORKFLOW_TEMPLATE_FOLDER_NAME;
	public static final String WORKFLOW_TASK_PATH = WORKFLOW_FOLDER_PATH +"/"+ WORKFLOW_TASK_FOLDER_NAME;
	public static final String WORKFLOW_TASK_DEFINITION_PATH = WORKFLOW_FOLDER_PATH + "/" + WORKFLOW_TASK_DEFINITION_FOLDER_NAME;

	public static final String PROCSTATE_TASK_TODO = "todo";
	public static final String PROCSTATE_TASK_DONE = "done";
	public static final String PROCSTATE_TRANSITION_READY = "transition_ready";
	public static final String PROCSTATE_WORKFLOW_STARTED = "running";
	public static final String PROCSTATE_WORKFLOW_FINISHED = "finished";
	public static final String PROCSTATE_REVIEW_OK = "review_ok";

    // Constants for the RenderServer extension
    public static final String RENDER_SERVER_LIFECYCLE = "_RenderServerLC";
    public static final String RENDERSERVER_RENDER_TASK_NEW = "newRenderTask";
    public static final String RENDERSERVER_RENDER_TASK_RENDERING = "executeRenderTask";
    public static final String RENDERSERVER_RENDER_TASK_FINISHED = "finishedRenderTask";
    public static final String RENDERSERVER_RENDER_TASK_FAILED = "failedRenderTask";
    public static final String OBJECT_TYPE_RENDER_TASK = "_render_task";

    public static final int LARGE_TEXT_FIELD = 10485760;
    // max_field_size in postgres is 1 GByte, but somehow Hibernate creates a varchar instead of text and that is
    // limited to 10 MByte.

    public static final int METADATA_SIZE = LARGE_TEXT_FIELD;
    public static final int NAME_LENGTH = 128;

	public static final int DESCRIPTION_SIZE = 255;

	/**
	 * Default length of database fields which contain XPATH statements.
	 */
	public static final int XPATH_LENGTH = LARGE_TEXT_FIELD;
	
	/**
	 * Default length of database fields which contain XML parameters.
	 */
	public static final int XML_PARAMS = LARGE_TEXT_FIELD;

    /*
     * We cannot make the INDEX_* and FIELD_* into an enum, because the user
     * may add their own index items and fields dynamically. But the constants here
     * are to be used for code that needs to refer to system-relevant index items and fields.
     * Also, for historical reasons, the name of the index item is not always fully similar to the field name.
     */
    public static final String INDEX_FOLDER_OWNER = "index.folder.owner";
    public static final String FIELD_FOLDER_OWNER = "owner";
    public static final String INDEX_FOLDER_NAME = "index.folder.name";
    public static final String FIELD_FOLDER_NAME = "name";
    public static final String INDEX_FOLDER_TYPE = "index.folder.type";
    public static final String FIELD_FOLDER_TYPE = "folderType";
    public static final String INDEX_FOLDER_PARENT_ID = "index.folder.parent_id";
    public static final String FIELD_FOLDER_PARENT_ID = "parent_id";
    public static final String INDEX_FOLDER_PATH = "index.folder.path";
    public static final String FIELD_FOLDER_PATH = "folderpath";
    public static final String INDEX_PROCSTATE = "index.procstate";
    public static final String FIELD_PROCSTATE = "procstate";
    public static final String INDEX_APPNAME = "index.appname";
    public static final String FIELD_APPNAME = "appname";
    public static final String INDEX_VERSION = "index.version";
    public static final String FIELD_VERSION = "version";
    public static final String INDEX_NAME = "index.name";
    public static final String FIELD_NAME = "name";
    public static final String INDEX_OBJECT_TYPE_NAME = "index.object_type_name";
    public static final String FIELD_OBJECT_TYPE_NAME = "objectTypeName";
    public static final String INDEX_ROOT = "index.root";
    public static final String FIELD_ROOT = "root";
    public static final String INDEX_PARENT_ID = "index.parentId";
    public static final String FIELD_PARENT_ID = "parent_id";
    public static final String INDEX_CONTENT_SIZE = "index.contentsize";
    public static final String FIELD_CONTENT_SIZE = "contentsize";
    public static final String INDEX_LOCKED_BY = "index.lockedby";
    public static final String FIELD_LOCKED_BY = "lockedby";
    public static final String INDEX_CREATOR = "index.creator";
    public static final String FIELD_CREATOR = "creator";
    public static final String INDEX_MODIFIER = "index.modifier";
    public static final String FIELD_MODIFIER = "modifier";
    public static final String INDEX_OWNER = "index.owner";
    public static final String FIELD_OWNER = "owner";
    public static final String INDEX_FORMAT = "index.format";
    public static final String FIELD_FORMAT = "format";
    public static final String INDEX_OBJECT_TYPE = "index.objecttype";
    public static final String FIELD_OBJECT_TYPE = "objecttype";
    public static final String INDEX_LANGUAGE = "index.language";
    public static final String FIELD_LANGUAGE = "language";
    public static final String INDEX_ACL = "index.acl";
    public static final String FIELD_ACL = "acl";
    public static final String INDEX_LIFECYCLE_STATE = "index.lifecycle.state";
    public static final String FIELD_LIFECYCLE_STATE = "lifecyclestate";
    public static final String INDEX_CREATED_DATE = "index.created.date";
    public static final String FIELD_CREATED_DATE = "created.date";
    public static final String INDEX_MODIFIED_DATE = "index.modified.date";
    public static final String FIELD_MODIFIED_DATE = "modified.date";
    public static final String INDEX_CREATED_TIME = "index.created.time";
    public static final String FIELD_CREATED_TIME = "created.time";
    public static final String INDEX_MODIFIED_TIME = "index.modified.time";
    public static final String FIELD_MODIFIED_TIME = "modified.time";
    public static final String INDEX_LATEST_HEAD = "index.latesthead";
    public static final String FIELD_LATEST_HEAD = "latesthead";
    public static final String INDEX_LATEST_BRANCH = "index.latestbranch";
    public static final String FIELD_LATEST_BRANCH = "latestbranch";
    public static final String INDEX_PATH = "index.path";
    public static final String FIELD_PATH = "folderpath";

    public static void main(String... args){
        System.out.print(Constants.SERVER_VERSION);
    }
}


src/server/global/PermissionName.java
package server.global;

/**
 *
 */
public class PermissionName {

    /**
     * Permission to lock an object. Implies permission to unlock the user's own locks.
     */
    public static final String LOCK = "_lock";

    /**
     * Permission to write to an object, that is: change its content.
     */
    public static final String WRITE_OBJECT_CONTENT = "_write_object_content";

    /**
     * Permission to read to an object's content.
     */
    public static final String READ_OBJECT_CONTENT = "_read_object_content";

    /**
     * Permission to change an object's system metadata
     */
    public static final String WRITE_OBJECT_SYS_METADATA = "_write_object_sysmeta";

    /**
     * Permission to read an object's system metadata
     */
    public static final String READ_OBJECT_SYS_METADATA = "_read_object_sysmeta";

    /**
     * Permission to change an object's custom metadata
     */
    public static final String WRITE_OBJECT_CUSTOM_METADATA = "_write_object_custom_metadata";

    /**
     * Permission to read an object's custom metadata
     */
    public static final String READ_OBJECT_CUSTOM_METADATA = "_read_object_custom_metadata";

    /**
     * Permission to create a new version of an object.
     */
    public static final String VERSION_OBJECT = "_version";

    /**
     * Permission to delete an object.
     */
    public static final String DELETE_OBJECT = "_delete_object";

    /**
     * Permission to browse an object, that is: the server will list it when
     * queried for the content of a folder or when a search for turns up this
     * object.
     */
    public static final String BROWSE_OBJECT = "_browse";

    /**
     * Permission to create a folder inside the current one.
     */
    public static final String CREATE_FOLDER = "_create_folder";

    /**
     * Permission to delete a folder.
     */
    public static final String DELETE_FOLDER = "_delete_folder";

    /**
     * Permission to create an object inside a folder.
     */
    public static final String CREATE_OBJECT = "_create_inside_folder"; // create_object_inside_folder
    /**
     * Permission to edit a folder (change name, change metadata).
     */
    public static final String EDIT_FOLDER = "_edit_folder";

    /**
     * Permission to browse this folder, that is: the folder will be
     * displayed in a list of its parent's content or may turn up
     * during a search.
     */
    public static final String BROWSE_FOLDER = "_browse_folder";

    /**
     * Permission to move an object or folder, depending on which it is set.
     */
    public static final String MOVE = "_move";

    /**
     * Permission to change an object's or folder's ACL.
     */
    public static final String SET_ACL = "_set_acl";

    /**
     * Permission to execute queries against custom tables.
     */
    public static final String QUERY_CUSTOM_TABLE = "_query_custom_table";

    /**
     * Permission to execute customized queries against the OSD-table.
     * Currently not used.
     */
    public static final String EXECUTE_QUERY = "_execute_query";

    /**
     * Permission to create an instance of a template object of some kind.
     * Current use case is for workflow templates. If you have a create_instance
     * permission on a workflow template, you may call create_workflow to instantiate a new workflow.
     */
    public static final String CREATE_INSTANCE = "_create_instance";
}


src/server/exceptions/CinnamonConfigurationException.java
// cinnamon - the Open Enterprise CMS project
// Copyright (C) 2007-2009 Horner GmbH (http://www.horner-project.eu)
// 
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
// 
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
// (or visit: http://www.gnu.org/licenses/lgpl.html)

package server.exceptions;

public class CinnamonConfigurationException extends RuntimeException {

	/**
	 * 
	 */
	private static final long	serialVersionUID	= 1L;

	public CinnamonConfigurationException() {
		super();
	}

	public CinnamonConfigurationException(String arg0, Throwable arg1) {
		super(arg0, arg1);
	}

	public CinnamonConfigurationException(String arg0) {
		super(arg0);
	}

	public CinnamonConfigurationException(Throwable arg0) {
		super(arg0);
	}

}


src/server/exceptions/CinnamonException.java
// cinnamon - the Open Enterprise CMS project
// Copyright (C) 2007-2009 Horner GmbH (http://www.horner-project.eu)
// 
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
// 
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
// (or visit: http://www.gnu.org/licenses/lgpl.html)

package server.exceptions;

import org.dom4j.Element;
import server.interfaces.ILocalizer;

public class CinnamonException extends RuntimeException {

	/**
	 * 
	 */
	private static final long	serialVersionUID	= 1L;
	
	String[] params = {};
	
	public CinnamonException() {
		super();
	}
	
	public CinnamonException(String message, String... params){
		super(message);
		if(params != null && params.length > 0){
			this.params = params;
		}
	}

	public CinnamonException(String arg0, Throwable arg1) {
		super(arg0, arg1);
	}

    public CinnamonException(String message, Throwable exception, String... params) {
		super(message, exception);
        if(params != null && params.length > 0){
			this.params = params;
		}
	}

	public CinnamonException(String arg0) {
		super(arg0);
	}

	public CinnamonException(Throwable arg0) {
		super(arg0);
	}
	
	public String[] getParams(){
		return params;
	}
	
	public void addToElement(Element root, ILocalizer localizer){
        String localizedMessage = getMessage();
        if(localizer != null && localizer.getInitialized()){
            localizedMessage = localizer.localize(getMessage());
            if(localizedMessage == null){
                localizedMessage = "null ("+getStackTrace()[0]+")";
            }
        }
		root.addElement("message").addText(localizedMessage);
		root.addElement("code").addText(getMessage() == null ? "null" : getMessage());
		Element paramList = root.addElement("parameters");
		if(params.length > 0){
			for(String p : params){
                if(p == null){
                    p = "null";
                }
				paramList.addElement("param").addText(p);
			}
		}
	}
}
